zerver/data_import/slack.py

```diff
@@ -1120,14 +1120,29 @@ def build_reactions(
     for realm_emoji in zerver_realmemoji:
         realmemoji[realm_emoji["name"]] = realm_emoji["id"]
 
+    # Slack's data exports use encode skin tone variants on emoji
+    # reactions like this: `clap::skin-tone-2`. For now, we only
+    # use the name of the base emoji, since Zulip's emoji
+    # reactions system doesn't yet support skin tone modifiers.
+    # We need to merge and dedup reactions, as someone may have
+    # reacted to `clap::skin-tone-1` and `clap::skin-tone-2`, etc.
+    mergedreactions: Dict[str, Any] = {}
+    for slack_reaction in reactions:
+        emoji_name = slack_reaction["name"].split("::", maxsplit=1)[0]
+        if emoji_name in mergedreactions:
+            mergedreactions[emoji_name]["users"] = list(
+                set(mergedreactions[emoji_name]["users"] + slack_reaction["users"])
+            )
+            mergedreactions[emoji_name]["count"] = len(mergedreactions[emoji_name]["users"])
+        else:
+            slack_reaction["name"] = emoji_name
+            mergedreactions[emoji_name] = slack_reaction
+    reactions = list(mergedreactions.values())
```

Comment:
I think we can simplify this in a couple ways -- just keeping the `set` (the ordering doesn't matter, and we just iterate over `slack_reaction["users"]`, which will work fine with a set), and using a `defaultdict` to remove the need for a special-case `if emoji_name in mergedreactions` check.

This is totally untested, but something like:
```suggestion
    merged_reactions = defaultdict(set)
    for slack_reaction in reactions:
        emoji_name = slack_reaction["name"].split("::", maxsplit=1)[0]
        merged_reactions[emoji_name].add(slack_reaction["user"])
    reactions = [{"name": k, "users": v, "count": len(v)} for k, v in merged_reactions.items()]
```

That also saves us from computing the `len` each time, which can make this quadratic.

Reply:
This code doesn't compile. I made several changes to progressively clear out stackdumps, and eventually just moved the core ideas out to a PoC and after reducing it to the simplest form, it then doesn't result in a merged deduped list.

```
#Before
[
    {"name": "grinning", "users": ["U061A5N1G"], "count": 1},
    {"name": "+1", "users": ["U061A5N1G", "U061A1R2R"], "count": 2},
    {"name": "+1::skin-tone-2", "users": ["U061A5N1G", "U066MTL5U"], "count": 2},
]
# After
[
    {"name": "grinning", "users": ["U061A5N1G"], "count": 1},
    {"name": "+1", "users": ["U066MTL5U", "U061A1R2R"], "count": 2},
]
# Should be
[
    {"name": "grinning", "users": ["U061A5N1G"], "count": 1},
    {"name": "+1", "users": ["U066MTL5U", "U061A1R2R", "U061A5N1G"], "count": 3},
]
```

Reply:
> That also saves us from computing the len each time, which can make this quadratic.

That could be put off to the end without a problem. The count field is never used in the zulipification process that I've seen.

Reply:
It works in my case:
```python
from collections import defaultdict

reactions = [
    {"name": "grinning", "users": ["U061A5N1G"], "count": 1},
    {"name": "+1", "users": ["U061A5N1G", "U061A1R2R"], "count": 2},
    {"name": "+1::skin-tone-2", "users": ["U061A5N1G", "U066MTL5U"], "count": 2},
]

merged_reactions = defaultdict(set)
for slack_reaction in reactions:
    emoji_name = slack_reaction["name"].split("::", maxsplit=1)[0]
    merged_reactions[emoji_name].update(slack_reaction["users"])
reactions = [{"name": k, "users": v, "count": len(v)} for k, v in merged_reactions.items()]
print(reactions)
# [{'name': 'grinning', 'users': {'U061A5N1G'}, 'count': 1}, {'name': '+1', 'users': {'U061A5N1G', 'U061A1R2R', 'U066MTL5U'}, 'count': 3}]
```
I agree that the `count` field is not used.